java/src/main/native/ai_onnxruntime_OrtSession.c

```diff
@@ -13,227 +13,286 @@
-        // Read out the input name and convert it to a java.lang.String
-        char* inputName;
-        checkOrtStatus(jniEnv,api,api->SessionGetInputName((OrtSession*)sessionHandle, i, allocator, &inputName));
-        jstring name = (*jniEnv)->NewStringUTF(jniEnv,inputName);
-        checkOrtStatus(jniEnv,api,api->AllocatorFree(allocator,inputName));
-
-        // Create a ValueInfo from the OrtTypeInfo
-        OrtTypeInfo* typeInfo;
-        checkOrtStatus(jniEnv,api,api->SessionGetInputTypeInfo((OrtSession*)sessionHandle, i, &typeInfo));
-        jobject valueInfoJava = convertToValueInfo(jniEnv,api,typeInfo);
-        api->ReleaseTypeInfo(typeInfo);
-
-        // Create a NodeInfo and assign into the array
-        jobject nodeInfo = (*jniEnv)->NewObject(jniEnv, nodeInfoClazz, nodeInfoConstructor, name, valueInfoJava);
-        (*jniEnv)->SetObjectArrayElement(jniEnv, array,safecast_size_t_to_jsize(i),nodeInfo);
+JNIEXPORT jobjectArray JNICALL Java_ai_onnxruntime_OrtSession_getInputInfo(JNIEnv* jniEnv, jobject jobj, jlong apiHandle, jlong sessionHandle, jlong allocatorHandle) {
+  (void)jobj;  // Required JNI parameter not needed by functions which don't need to access their host object.
+  const OrtApi* api = (const OrtApi*)apiHandle;
+  OrtSession* session = (OrtSession*)sessionHandle;
+  OrtAllocator* allocator = (OrtAllocator*)allocatorHandle;
+
+  // Setup
+  char* nodeInfoClassName = "ai/onnxruntime/NodeInfo";
+  jclass nodeInfoClazz = (*jniEnv)->FindClass(jniEnv, nodeInfoClassName);
+  jmethodID nodeInfoConstructor = (*jniEnv)->GetMethodID(jniEnv, nodeInfoClazz, "<init>",
+                                                         "(Ljava/lang/String;Lai/onnxruntime/ValueInfo;)V");
+
+  // Get the number of inputs
+  size_t numInputs = 0;
+  OrtErrorCode code = checkOrtStatus(jniEnv, api, api->SessionGetInputCount(session, &numInputs));
+  if (code != ORT_OK) {
+    return NULL;
+  }
+
+  // Allocate the return array
+  jobjectArray array = (*jniEnv)->NewObjectArray(jniEnv, safecast_size_t_to_jsize(numInputs), nodeInfoClazz, NULL);
+  for (size_t i = 0; i < numInputs; i++) {
+    // Read out the input name and convert it to a java.lang.String
+    char* inputName = NULL;
+    code = checkOrtStatus(jniEnv, api, api->SessionGetInputName(session, i, allocator, &inputName));
+    if (code != ORT_OK) {
+      break;
+    }
+    jstring name = (*jniEnv)->NewStringUTF(jniEnv, inputName);
+    code = checkOrtStatus(jniEnv, api, api->AllocatorFree(allocator, inputName));
+    if (code != ORT_OK) {
+      break;
     }
 
-    return array;
+    // Create a ValueInfo from the OrtTypeInfo
+    OrtTypeInfo* typeInfo = NULL;
+    code = checkOrtStatus(jniEnv, api, api->SessionGetInputTypeInfo(session, i, &typeInfo));
+    if (code != ORT_OK) {
+      break;
+    }
+    jobject valueInfoJava = convertToValueInfo(jniEnv, api, typeInfo);
+    api->ReleaseTypeInfo(typeInfo);
+    if (valueInfoJava == NULL) {
+      break;
+    }
```

Comment:
This feels to be an error. We need a way to report it?

Reply:
If it's NULL then the method which creates the value info has already thrown an exception into Java with the error message, all we need to do here is clean things up and return to the JVM.